File: Python/OneLang/One/Transforms/FillParent.py
from OneLangStdLib import *
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Statements as stats
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.AstTransformer as astTrans

class FillParent(astTrans.AstTransformer):
    def __init__(self):
        self.parent_node_stack = []
        super().__init__("FillParent")
    
    def visit_expression(self, expr):
        if len(self.parent_node_stack) == 0:
            pass
        expr.parent_node = self.parent_node_stack[len(self.parent_node_stack) - 1]
        self.parent_node_stack.append(expr)
        super().visit_expression(expr)
        self.parent_node_stack.pop()
        return expr
    
    def visit_statement(self, stmt):
        self.parent_node_stack.append(stmt)
        super().visit_statement(stmt)
        self.parent_node_stack.pop()
        return stmt
    
    def visit_enum(self, enum_):
        enum_.parent_file = self.current_file
        super().visit_enum(enum_)
        for value in enum_.values:
            value.parent_enum = enum_
    
    def visit_interface(self, intf):
        intf.parent_file = self.current_file
        super().visit_interface(intf)
    
    def visit_class(self, cls_):
        cls_.parent_file = self.current_file
        super().visit_class(cls_)
    
    def visit_global_function(self, func):
        func.parent_file = self.current_file
        super().visit_global_function(func)
    
    def visit_field(self, field):
        field.parent_interface = self.current_interface
        
        self.parent_node_stack.append(field)
        super().visit_field(field)
        self.parent_node_stack.pop()
    
    def visit_property(self, prop):
        prop.parent_class = self.current_interface
        
        self.parent_node_stack.append(prop)
        super().visit_property(prop)
        self.parent_node_stack.pop()
    
    def visit_method_base(self, method):
        if isinstance(method, types.Constructor):
            method.parent_class = self.current_interface
        elif isinstance(method, types.Method):
            method.parent_interface = self.current_interface
        elif isinstance(method, types.GlobalFunction):
            pass
        elif isinstance(method, types.Lambda):
            pass
        else:
            pass
        
        for param in method.parameters:
            param.parent_method = method
        
        self.parent_node_stack.append(method)
        super().visit_method_base(method)
        self.parent_node_stack.pop()
    
    def visit_file(self, file):
        for imp in file.imports:
            imp.parent_file = file
        
        super().visit_file(file)

File: PHP/src/OneLang/One/Transforms/FillParent.php
<?php

namespace OneLang\One\Transforms\FillParent;

use OneLang\One\Ast\Types\SourceFile;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Types\IInterface;
use OneLang\One\Ast\Types\Enum;
use OneLang\One\Ast\Types\Interface_;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Field;
use OneLang\One\Ast\Types\Property;
use OneLang\One\Ast\Types\IAstNode;
use OneLang\One\Ast\Types\IMethodBase;
use OneLang\One\Ast\Types\Constructor;
use OneLang\One\Ast\Types\GlobalFunction;
use OneLang\One\Ast\Types\Lambda;
use OneLang\One\Ast\Statements\Statement;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\AstTransformer\AstTransformer;

class FillParent extends AstTransformer {
    public $parentNodeStack;
    
    function __construct() {
        parent::__construct("FillParent");
        $this->parentNodeStack = array();
    }
    
    protected function visitExpression($expr) {
        if (count($this->parentNodeStack) === 0) { }
        $expr->parentNode = $this->parentNodeStack[count($this->parentNodeStack) - 1];
        $this->parentNodeStack[] = $expr;
        parent::visitExpression($expr);
        array_pop($this->parentNodeStack);
        return $expr;
    }
    
    protected function visitStatement($stmt) {
        $this->parentNodeStack[] = $stmt;
        parent::visitStatement($stmt);
        array_pop($this->parentNodeStack);
        return $stmt;
    }
    
    protected function visitEnum($enum_) {
        $enum_->parentFile = $this->currentFile;
        parent::visitEnum($enum_);
        foreach ($enum_->values as $value)
            $value->parentEnum = $enum_;
    }
    
    protected function visitInterface($intf) {
        $intf->parentFile = $this->currentFile;
        parent::visitInterface($intf);
    }
    
    protected function visitClass($cls) {
        $cls->parentFile = $this->currentFile;
        parent::visitClass($cls);
    }
    
    protected function visitGlobalFunction($func) {
        $func->parentFile = $this->currentFile;
        parent::visitGlobalFunction($func);
    }
    
    protected function visitField($field) {
        $field->parentInterface = $this->currentInterface;
        
        $this->parentNodeStack[] = $field;
        parent::visitField($field);
        array_pop($this->parentNodeStack);
    }
    
    protected function visitProperty($prop) {
        $prop->parentClass = $this->currentInterface;
        
        $this->parentNodeStack[] = $prop;
        parent::visitProperty($prop);
        array_pop($this->parentNodeStack);
    }
    
    protected function visitMethodBase($method) {
        if ($method instanceof Constructor)
            $method->parentClass = $this->currentInterface;
        else if ($method instanceof Method)
            $method->parentInterface = $this->currentInterface;
        else if ($method instanceof GlobalFunction) { }
        else if ($method instanceof Lambda) { }
        else { }
        
        foreach ($method->parameters as $param)
            $param->parentMethod = $method;
        
        $this->parentNodeStack[] = $method;
        parent::visitMethodBase($method);
        array_pop($this->parentNodeStack);
    }
    
    function visitFile($file) {
        foreach ($file->imports as $imp)
            $imp->parentFile = $file;
        
        parent::visitFile($file);
    }
}
